[Alternate WebM Player] Incorrect first frame paint

youssefsoli opened the pull request:
#### 14204636ed8ad06e4d89b952f40b174d35a57f6f
<pre>
[Alternate WebM Player] Incorrect first frame paint
<a href="https://bugs.webkit.org/show_bug.cgi?id=243281">https://bugs.webkit.org/show_bug.cgi?id=243281</a>
&lt;rdar://97701232&gt;

Reviewed by Eric Carlson.

Currently there is a race condition when attempting to paint the first
video frame on builds that do not support AVSBDL -copyDisplayedPixelBuffer
directly related to the cause of <a href="https://bugs.webkit.org/show_bug.cgi?id=228997.">https://bugs.webkit.org/show_bug.cgi?id=228997.</a>

Due to the asynchronous nature of the WebCoreDecompressionSession and the
very short time between destroying the AVSBDL and creating a new decompression
session, we run into a scenario where the first video frame is requested by
WebGLRenderingContextBase::videoFrameToImage is not available causing us
to draw an empty black frame.

This also affected the legacy code path since the first frame sometimes
wasn&apos;t available for the call to `colorSpace` but was available for the
subsequent call to paint, resulting in a frame being painted with the
incorrect color space.

The fix is to make the already synchronous `nativeImageForCurrentTime`
call even more synchronous on the private player layer and wait for the
first frame to be decoded before attempting to fetch it.

This kills any chance of the legacy code path from being taken so there
is no need to make those calls synchronous as well.

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::videoFrameForCurrentTime):
(WebCore::MediaPlayerPrivateWebM::flushVideo):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::registerNotifyWhenHasAvailableVideoFrame):

Canonical link: <a href="https://commits.webkit.org/253059@main">https://commits.webkit.org/253059@main</a>
</pre>

youssefsoli commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1353,6 +1371,10 @@ static bool isCopyDisplayedPixelBufferAvailable()
     m_decompressionSession->invalidate();
     m_decompressionSession = nullptr;
+    if (m_hasAvailableVideoFrameSemaphore) {
+        m_hasAvailableVideoFrameSemaphore->signal();
```
Not sure if I should keep this `signal()` call considering this is being called on the main thread (which would be blocked regardless)

eric-carlson replied:
I don't think it is possible, but why don't you  `ASSERT(! m_hasAvailableVideoFrameSemaphore)`

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -1261,8 +1276,13 @@ static bool isCopyDisplayedPixelBufferAvailable()
+                m_hasAvailableVideoFrameSemaphore = nullptr;
+            }
+        }
     });
```
This is repeated in `flushVideo`. While it isn't a lot of code, it would be better to share the code if possible.

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -484,8 +484,14 @@ static bool isCopyDisplayedPixelBufferAvailable()
+        // work around https://bugs.webkit.org/show_bug.cgi?id=228997
+        // on builds without AVSAMPLEBUFFERDISPLAYLAYER_COPYDISPLAYEDPIXELBUFFER
+        if (m_hasAvailableVideoFrameSemaphore && m_decompressionSession)
+            m_hasAvailableVideoFrameSemaphore->waitFor(100_ms);
```
It would be better to make this a named constant.

eric-carlson reviewed (approved):
r=me

youssefsoli committed:
[Alternate WebM Player] Incorrect first frame paint
https://bugs.webkit.org/show_bug.cgi?id=243281
<rdar://97701232>

Reviewed by Eric Carlson.

Currently there is a race condition when attempting to paint the first
video frame on builds that do not support AVSBDL -copyDisplayedPixelBuffer
directly related to the cause of https://bugs.webkit.org/show_bug.cgi?id=228997.

Due to the asynchronous nature of the WebCoreDecompressionSession and the
very short time between destroying the AVSBDL and creating a new decompression
session, we run into a scenario where the first video frame is requested by
WebGLRenderingContextBase::videoFrameToImage is not available causing us
to draw an empty black frame.

This also affected the legacy code path since the first frame sometimes
wasn't available for the call to `colorSpace` but was available for the
subsequent call to paint, resulting in a frame being painted with the
incorrect color space.

The fix is to make the already synchronous `nativeImageForCurrentTime`
call even more synchronous on the private player layer and wait for the
first frame to be decoded before attempting to fetch it.

This kills any chance of the legacy code path from being taken so there
is no need to make those calls synchronous as well.

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::videoFrameForCurrentTime):
(WebCore::MediaPlayerPrivateWebM::flushVideo):
(WebCore::MediaPlayerPrivateWebM::ensureDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::destroyDecompressionSession):
(WebCore::MediaPlayerPrivateWebM::registerNotifyWhenHasAvailableVideoFrame):

Canonical link: https://commits.webkit.org/253059@main
